### ISSUE TITLE
Higher-kinds

### ISSUE THREAD

**Issue by shelby3** [keean/zenscript#10]
Kinds are [the type of type constructors](https://en.wikipedia.org/wiki/Kind_(type_theory)). Thus higher-kinds are constructors on type constructors.

Some [prior discussion at the Rust forum](https://users.rust-lang.org/t/does-rust-really-need-higher-kinded-types/5531/8) about higher-kinds may provide a clearer understanding of what they are and use cases.

It is important to note that the supertype of a union contains more types than its subtype, e.g. `Number|String` is the supertype of `Number`. Thus we can not subsume to the greatest lower bound (aka GLB or common supertype) of the element type of an instance of an `Array`, e.g. from `Number` to `Number|String`, because some preexisting reference of type `Array<Number>` would access a data structure that contains also `String` elements. Thus such a subsumption will not type check. Thus we are unable to add or insert elements (into an instance of `Array`) which have a type that is not already the subtype of the preexisting element type. Thus employing `Array` with unions is not a solution to the Expression Problem, because we can add new variants (types). Tangentially note that subsuming to the superclass for subclassing or typeclass objects (`trait` in Rust) works fine with `Array`, because the dynamic dispatch insures that no preexisting references to the `Array` instance are invalid by inserting or adding elements with new subtypes to the data structure, and thus it type checks. But subclassing and typeclass objects do not solve the Expression Problem, because new operations can't be added (for subclassing can't be added after instantiation of the class and for typeclass objects can't be added after instantiation of the typeclass object).

But a `List` doesn't have this problem and is an invariant data structure, because the `head` of the `Cons` can be any type ([if we are subsuming to a new union type](https://github.com/keean/zenscript/issues/8#issuecomment-249049702)) and any preexisting references to any element in the tail of the list will never access the new head, because a `List` is a singly-linked data structure with new elements prepended. Thus lists allow adding new variants (types) even for union types (not just for subclassing and typeclass objects) and since this retains the explicit union type instead of premature subsumption to a dynamic dispatch (employed by subclassing and typeclass objects), then we can also add new operations as a delayed binding to the typeclass bound. This was my invention as a solution to the Expression Problem.

When [I started thinking about](https://github.com/keean/zenscript/issues/8#issuecomment-248954012) how a `List` would interact with the first-class unions we accepted into ZenScript's design, I realized that [if typeclasses can't provide a factory](https://github.com/keean/zenscript/issues/8#issuecomment-249024325) to construct an new instance of the data structure they are operating on, then [it is impossible to write some generic operations such as a sort](https://github.com/keean/zenscript/issues/8#issuecomment-249039330) when employing subsuming heterogeneous unions with higher-order containers.

For example, a sort operation needs to re-order the elements of a collection.

(more to be added, this OP is still being composed)

**Comment by shelby3** [comment]
Done. Feel free to add comments.

**Comment by shelby3** [comment]
It appears we can emit [TypeScript code to implement](https://github.com/keean/zenscript/issues/13#issuecomment-249399737) `Monoid` with my hack without needing higher-kinds, but what syntax will that have in ZenScript?

```
typeclass Monoid<T<A>> { // A is inferred to be a type parameter because ALLCAPS
    identity(): T<A>
    append(A): T<A>
}
List<A> implements Monoid<List<A>> {
    identity() => Nil
    append(x) => Cons(x, this)
}
```

Since `List<A>` is not a subsclass of `Monoid` then apparently there is no higher-kinds involved, because the compiler doesn't have to reason about the constructor `T<>` in the context of the constructor `Monoid<>`. Is that correct?

**Comment by keean** [comment]
I don't think that is quite right. Monoid takes a parameter in no-object systems because is declaring free functions not methods:
```
data List<T> = Nil | Cons{head: T, tail: List<T>}
typeclass Monoid<A> { 
    identity(): A
    append(A, A): A
}
implement Monoid<List<A>> {
    identity() => Nil
    append(x) => Cons(x, this)
}
```
The above is how I would want to do it in a no-objects system. The following is for an object system:
```
data List<T> = Nil | Cons{head: T, tail: List<T>}
typeclass Monoid {
    identity(): This
    append(This): This
}
List<A> implements Monoid {
    identity() => Nil
    append(x) => Cons(x, this)
}
```

**Comment by keean** [comment]
There is a further small problem with the above, identity is effectively a constructor function, so the second example is not quite right. Infact we would have to introduce a new keyword to make `identity` a static function in C++ terminology. The first example (non-objects) is correct though.

**Comment by shelby3** [comment]
@keean wrote:
> The above is how I would want to do it in a no-objects system.

Your `append` doesn't make any sense. The `x` has to have the type of the elements of `List<A>`.

**Comment by keean** [comment]
You are right, my implementation is wrong, but the type signature is correct, I was too quick cut-and-pasting from the posted code:
```
append(x, y) => match(x):
    Nil -> return y
    Cons(h, t) -> Cons(h, append(t, y))
```

**Comment by shelby3** [comment]
@keean wrote:
> ```
> append(x, y) => match(x):
>     Nil -> return y
>     Cons(h, t) -> Cons(h, append(t, y))
> ```

Yours is apparently the category theory correct Monoid and mine wasn't, but I don't see how you can implement the `sort` [generically for invariant](https://github.com/keean/zenscript/issues/8#issuecomment-249039330) `List` with the above? Will we have another typeclass interface that can return any range (by iterators) of itself? Or do we just use a clone typeclass to get a copy and sort in place on that? But sort in place (generically) on a list is very inefficient.

**Comment by shelby3** [comment]
refresh

**Comment by shelby3** [comment]
@keean wrote:
> Infact we would have to introduce a new keyword to make identity a static function in C++ terminology.

I showed how to do it in TypeScript using only type parameters. Perhaps you mean without the unsound runtime exceptions I had to use?

**Comment by shelby3** [comment]
Note the edit:

> Edit: seems the most efficient way to sort a list is to convert it to an array, sort, then convert it back to list. So how to do that generically?

**Comment by shelby3** [comment]
refresh

**Comment by keean** [comment]
To efficiently sort a list, try a merge-sort. Recursively split the list into two (can be done with references to the source list without changing it). When you get to single items take each item and put into a separate list. Then perform a merge operation on each pair of lists, repeatedly until only one list remains which will be sorted.

The merge operation uses three lists, to input and one output. whilst there is an element in both source lists, it takes the lowest head of the two (or highest depending on desired order) and pushes it into the output list removing it from the input list. This repeats until one or both input lists are empty. If both are empty the merge is finished, otherwise we push the remaining items into the output removing them from the input, as they will already be in order.

**Comment by shelby3** [comment]
Trade-off of [prematurely specializing](https://github.com/keean/zenscript/issues/22#issuecomment-251013757) the algorithm?

**Comment by shelby3** [comment]
Cross-referencing …

@shelby3 [wrote](https://github.com/keean/zenscript/issues/27#issuecomment-302019116):
> …
> 
> #### Higher-kinded Types
> 
> [Isomorphic](https://github.com/keean/zenscript/issues/27#issuecomment-252587142) to distilling the higher-_ranked_ issues, higher-_kinded_ types are higher-order type constructors where the ‘kind’ means the “type” (aka meta-type) of a type constructor with [meta-type variables](http://stackoverflow.com/questions/3034264/haskell-weird-kinds-kind-of-is), e.g. kind of order 0 `*` for primitive types such as `Int`, kind of order 1 `* → *` for `List[T]` and `* → * → *` for `Either[A, B]`, and kind of order 2 `(* → *) → *` for `Functor[List[T]]` and `(* → * → *) → *` for `Functor[Either[A, B]]`. Higher-kinded types are required to generically (type safely) abstract over generics such as for [Functor.map](http://stackoverflow.com/questions/21170493/when-are-higher-kinded-types-useful/21174565/#answer-21174565) ([also](https://users.rust-lang.org/t/does-rust-really-need-higher-kinded-types/5531/9?u=shelby3)) and [iteration](https://users.rust-lang.org/t/does-rust-really-need-higher-kinded-types/5531/11?u=shelby3).

**Comment by shelby3** [comment]
We [must support HKT](https://github.com/keean/zenscript/issues/27#issuecomment-402699154).
